File: src/main/java/tw/racket/lang/RacketSyntaxHighlighter.java
package tw.racket.lang;

import com.intellij.lexer.Lexer;
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors;
import com.intellij.openapi.editor.HighlighterColors;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase;
import com.intellij.psi.TokenType;
import com.intellij.psi.tree.IElementType;
import org.jetbrains.annotations.NotNull;
import tw.racket.lang.lexer.RacketLexerAdapter;
import tw.racket.lang.psi.RacketTypes;

import static com.intellij.openapi.editor.colors.TextAttributesKey.createTextAttributesKey;

public class RacketSyntaxHighlighter extends SyntaxHighlighterBase {
    @NotNull
    @Override
    public Lexer getHighlightingLexer() {
        return new RacketLexerAdapter();
    }

    @NotNull
    @Override
    public TextAttributesKey[] getTokenHighlights(IElementType tokenType) {
        if (tokenType.equals(RacketTypes.STRING_LITERAL)) {
            return STRING_KEYS;
        } else if (tokenType.equals(RacketTypes.NUMBER_LITERAL)) {
            return NUMBER_KEYS;
        } else if (tokenType.equals(RacketTypes.CHAR_LITERAL)) {
            return STRING_KEYS;
        } else if (tokenType.equals(RacketTypes.KEYWORD)) {
            return KEYWORD_KEYS;
        } else if (tokenType.equals(RacketTypes.COMMENT)) {
            return COMMENT_KEYS;
        } else if (tokenType.equals(TokenType.BAD_CHARACTER)) {
            return BAD_CHAR_KEYS;
        } else {
            return EMPTY_KEYS;
        }
    }

    public static final TextAttributesKey STRING =
            createTextAttributesKey("RACKET_STRING", DefaultLanguageHighlighterColors.STRING);
    public static final TextAttributesKey NUMBER =
            createTextAttributesKey("RACKET_NUMBER", DefaultLanguageHighlighterColors.NUMBER);
    public static final TextAttributesKey KEYWORD =
            createTextAttributesKey("RACKET_KEYWORD", DefaultLanguageHighlighterColors.KEYWORD);
    public static final TextAttributesKey COMMENT =
            createTextAttributesKey("RACKET_COMMENT", DefaultLanguageHighlighterColors.LINE_COMMENT);
    public static final TextAttributesKey BAD_CHARACTER =
            createTextAttributesKey("RACKET_BAD_CHARACTER", HighlighterColors.BAD_CHARACTER);


    private static final TextAttributesKey[] STRING_KEYS = new TextAttributesKey[]{STRING};
    private static final TextAttributesKey[] NUMBER_KEYS = new TextAttributesKey[]{NUMBER};
    private static final TextAttributesKey[] BAD_CHAR_KEYS = new TextAttributesKey[]{BAD_CHARACTER};
    private static final TextAttributesKey[] KEYWORD_KEYS = new TextAttributesKey[]{KEYWORD};
    private static final TextAttributesKey[] COMMENT_KEYS = new TextAttributesKey[]{COMMENT};
    private static final TextAttributesKey[] EMPTY_KEYS = new TextAttributesKey[0];
}

File: src/main/kotlin/tw/racket/lang/RacketSyntaxHighlighter.kt
package tw.racket.lang

import com.intellij.lexer.Lexer
import com.intellij.openapi.editor.DefaultLanguageHighlighterColors
import com.intellij.openapi.editor.HighlighterColors
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.fileTypes.SyntaxHighlighterBase
import com.intellij.psi.TokenType
import com.intellij.psi.tree.IElementType
import tw.racket.lang.lexer.RacketLexerAdapter
import tw.racket.lang.psi.RacketTypes

class RacketSyntaxHighlighter : SyntaxHighlighterBase() {
    override fun getHighlightingLexer(): Lexer {
        return RacketLexerAdapter()
    }

    override fun getTokenHighlights(tokenType: IElementType): Array<TextAttributesKey> {
        return if (tokenType == RacketTypes.STRING_LITERAL) {
            STRING_KEYS
        } else if (tokenType == RacketTypes.NUMBER_LITERAL) {
            NUMBER_KEYS
        } else if (tokenType == RacketTypes.CHAR_LITERAL) {
            STRING_KEYS
        } else if (tokenType == RacketTypes.COMMENT) {
            COMMENT_KEYS
        } else if (tokenType == RacketTypes.TRUE || tokenType == RacketTypes.FALSE) {
            BOOLEAN_KEYS
        } else if (tokenType == RacketTypes.MODULE_PLUS || tokenType == RacketTypes.PROVIDE || tokenType == RacketTypes.DEFINE || tokenType == RacketTypes.BEGIN || tokenType == RacketTypes.LET || tokenType == RacketTypes.LET_STAR || tokenType == RacketTypes.LET_REC || tokenType == RacketTypes.STRUCT || tokenType == RacketTypes.COND || tokenType == RacketTypes.MATCH) {
            KEYWORD_KEYS
        } else if (tokenType == TokenType.BAD_CHARACTER) {
            BAD_CHAR_KEYS
        } else if (tokenType == RacketTypes.LPAREN || tokenType == RacketTypes.RPAREN) {
            PARENTHESES_KEYS
        } else if (tokenType == RacketTypes.LBRACE || tokenType == RacketTypes.RBRACE) {
            BRACES_KEYS
        } else {
            arrayOf()
        }
    }

    companion object {
        private val BRACES_KEYS = arrayOf(TextAttributesKey.createTextAttributesKey("RACKET_BRACES", DefaultLanguageHighlighterColors.BRACES))
        private val PARENTHESES_KEYS = arrayOf(TextAttributesKey.createTextAttributesKey("RACKET_PARENTHESES", DefaultLanguageHighlighterColors.PARENTHESES))
        private val KEYWORD_KEYS = arrayOf(TextAttributesKey.createTextAttributesKey("RACKET_KEYWORD", DefaultLanguageHighlighterColors.KEYWORD))
        private val STRING_KEYS = arrayOf(TextAttributesKey.createTextAttributesKey("RACKET_STRING", DefaultLanguageHighlighterColors.STRING))
        private val NUMBER_KEYS = arrayOf(TextAttributesKey.createTextAttributesKey("RACKET_NUMBER", DefaultLanguageHighlighterColors.NUMBER))
        private val BOOLEAN_KEYS = arrayOf(TextAttributesKey.createTextAttributesKey("RACKET_BOOLEAN", DefaultLanguageHighlighterColors.CONSTANT))
        private val BAD_CHAR_KEYS = arrayOf(TextAttributesKey.createTextAttributesKey("RACKET_BAD_CHARACTER", HighlighterColors.BAD_CHARACTER))
        private val COMMENT_KEYS = arrayOf(TextAttributesKey.createTextAttributesKey("RACKET_COMMENT", DefaultLanguageHighlighterColors.LINE_COMMENT))
    }
}